llvm/test/Transforms/Util/loop-iter-count-assumptions.ll

```diff
@@ -0,0 +1,481 @@
+; CHECK-NEXT:    [[TMP1:%.*]] = load i32, ptr [[ARRAYIDX]], align 4
+; CHECK-NEXT:    [[ADD:%.*]] = add nsw i32 [[TMP1]], 8
+; CHECK-NEXT:    store i32 [[ADD]], ptr [[ARRAYIDX]], align 4
+; CHECK-NEXT:    [[DEC]] = add nsw i32 [[I_0]], -1
+; CHECK-NEXT:    [[TMP2:%.*]] = icmp sgt i32 [[TMP0]], 0
+; CHECK-NEXT:    call void @llvm.assume(i1 [[TMP2]])
+; CHECK-NEXT:    [[CMP:%.*]] = icmp sgt i32 [[DEC]], 0
+; CHECK-NEXT:    br i1 [[CMP]], label [[DO_BODY]], label [[DO_END:%.*]], !llvm.loop [[LOOP0]]
+; CHECK:       do.end:
+; CHECK-NEXT:    ret void
+;
+entry:
+  br label %do.body
+
+do.body:                                          ; preds = %do.body, %entry
+  %i.0 = phi i32 [ %n, %entry ], [ %dec, %do.body ]
+  %arrayidx = getelementptr inbounds i32, ptr %ptr, i32 %i.0
+  %1 = load i32, ptr %arrayidx, align 4
+  %add = add nsw i32 %1, 8
+  store i32 %add, ptr %arrayidx, align 4
+  %dec = add nsw i32 %i.0, -1
+  %cmp = icmp sgt i32 %dec, 0
+  br i1 %cmp, label %do.body, label %do.end, !llvm.loop !6
+
+do.end:                                           ; preds = %do.body
+  ret void
+}
+
+; guarded loop
+; if (n > 0) {
+;   for (i =0; i < n ; i++)
+; }
+; Function Attrs: mustprogress nounwind
+define dso_local void @guardedLoop(ptr %ptr, i32 noundef %n) #0 {
+; CHECK-LABEL: @guardedLoop(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[CMP:%.*]] = icmp sgt i32 [[N:%.*]], 0
+; CHECK-NEXT:    br i1 [[CMP]], label [[IF_THEN:%.*]], label [[IF_END:%.*]]
+; CHECK:       if.then:
+; CHECK-NEXT:    br label [[FOR_COND:%.*]]
+; CHECK:       for.cond:
+; CHECK-NEXT:    [[I_0:%.*]] = phi i32 [ 0, [[IF_THEN]] ], [ [[INC:%.*]], [[FOR_BODY:%.*]] ]
+; CHECK-NEXT:    [[TMP0:%.*]] = icmp slt i32 4, [[N]]
+; CHECK-NEXT:    call void @llvm.assume(i1 [[TMP0]])
+; CHECK-NEXT:    [[CMP1:%.*]] = icmp slt i32 [[I_0]], [[N]]
+; CHECK-NEXT:    br i1 [[CMP1]], label [[FOR_BODY]], label [[FOR_COND_CLEANUP:%.*]]
+; CHECK:       for.cond.cleanup:
+; CHECK-NEXT:    br label [[IF_END]]
+; CHECK:       for.body:
+; CHECK-NEXT:    [[INC]] = add nsw i32 [[I_0]], 1
+; CHECK-NEXT:    br label [[FOR_COND]], !llvm.loop [[LOOP0]]
+; CHECK:       if.end:
+; CHECK-NEXT:    ret void
+;
+entry:
+  %cmp = icmp sgt i32 %n, 0
+  br i1 %cmp, label %if.then, label %if.end
+
+if.then:                                          ; preds = %entry
+  br label %for.cond
+
+for.cond:                                         ; preds = %for.body, %if.then
+  %i.0 = phi i32 [ 0, %if.then ], [ %inc, %for.body ]
+  %cmp1 = icmp slt i32 %i.0, %n
+  br i1 %cmp1, label %for.body, label %for.cond.cleanup
+
+for.cond.cleanup:                                 ; preds = %for.cond
+  br label %if.end
+
+for.body:                                         ; preds = %for.cond
+  %inc = add nsw i32 %i.0, 1
+  br label %for.cond, !llvm.loop !6
+
+if.end:                                           ; preds = %for.cond.cleanup, %entry
+  ret void
+}
```

Comment:
Could you add a test case where there are two exiting blocks? E.g.
```
for (i = start_val; i < upper_b; ++i) {
if (i > lower_b) break;
do_something;
}
```

And maybe a case where both parts of the compare are IV:
```
for (i = 1; n = 20; i < n; i=2*i, ++n) { ... }
```